native_client/dotnet/DeepSpeechClient/DeepSpeech.cs

```diff
@@ -144,41 +162,88 @@ public unsafe void Dispose()
+        public unsafe void FeedAudioContent(DeepSpeechStream stream, short[] aBuffer, uint aBufferSize)
         {
-            NativeImp.DS_FeedAudioContent(_streamingStatePP, aBuffer, aBufferSize);
+            CheckAcousticModel();
+            CheckActiveStream(stream);
+            NativeImp.DS_FeedAudioContent(stream.StreamingStatePP, aBuffer, aBufferSize);
         }
 
         /// <summary>
         /// Closes the ongoing streaming inference, returns the STT result over the whole audio signal.
         /// </summary>
-        /// <returns>The STT result. The user is responsible for freeing the string.</returns>
-        public unsafe string FinishStream()
+        /// <param name="stream">Instance of the stream to finish.</param>
+        /// <returns>The STT result.</returns>
+        /// <exception cref="InvalidOperationException">Thrown when the acoustic model hasn't been initialized <see cref="CreateModel(string, uint)"/>.</exception> 
+        public unsafe string FinishStream(DeepSpeechStream stream)
         {
-            return NativeImp.DS_FinishStream(_streamingStatePP).PtrToString();
+            CheckAcousticModel();
+            CheckActiveStream(stream);
+            string finishStreamResult = NativeImp.DS_FinishStream(stream.StreamingStatePP).PtrToString();
+            stream.Active = false;
+            return finishStreamResult;
         }
 
         /// <summary>
         /// Closes the ongoing streaming inference, returns the STT result over the whole audio signal.
         /// </summary>
-        /// <returns>The extended metadata. The user is responsible for freeing the struct.</returns>
-        public unsafe Models.Metadata FinishStreamWithMetadata()
+        /// <param name="stream">Instance of the stream to finish.</param>
+        /// <returns>The extended metadata result.</returns>
+        /// <exception cref="InvalidOperationException">Thrown when the acoustic model hasn't been initialized <see cref="CreateModel(string, uint)"/> or passing an inactive stream.</exception>
+        public unsafe Metadata FinishStreamWithMetadata(DeepSpeechStream stream)
         {
-            return NativeImp.DS_FinishStreamWithMetadata(_streamingStatePP).PtrToMetadata();
+            CheckAcousticModel();
+            CheckActiveStream(stream);
+            var metadataResult = NativeImp.DS_FinishStreamWithMetadata(stream.StreamingStatePP).PtrToMetadata();
+            stream.Active = false;
+            return metadataResult;
+        }
+
+        /// <summary>
+        /// Checks if the stream is active.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if the stream is not active.</exception>
+        private void CheckActiveStream(DeepSpeechStream stream, [CallerMemberName]string callerMemberName = null)
```

Comment:
Same, I don't get why you would need this.

Reply:
Avoiding `System.AccessViolationException` makes sense? If the user makes a mistake there will be no clue about the actual issue. 

If we remove this check I will add warnings about it to the doc.

Now makes sense or you still want to remove it?

Reply:
> Avoiding `System.AccessViolationException` makes sense? If the user makes a mistake there will be no clue about the actual issue.
> 
> If we remove this check I will add warnings about it to the doc.
> 
> Now makes sense or you still want to remove it?

I still fear it's a bit over-engineered. Honestly, I'm not choked by throwing a `System.AccessViolationException`, but if you think it's needed, let's keep it. We can still add that in the doc, at the same time.